File: LTNC/Exercise/week_2/exercise_1/exercise_1.java
class Vehicle {
    private String brand;
    private int years;
    
    public Vehicle(){
        this.brand = "Unknow";
        this.years = 0;
    }
    
    public Vehicle(String brd, int yrs){
        this.brand = brd;
        this.years = yrs;
    }
    
    public void setBrand(String brd){
        this.brand = brd;
    }
    
    public void setYears(int y){
        this.years = y;
    }
    
    public String getBrand(){
        return this.brand;
    }
    public int getYears(){
        return this.years;
    }
    
    public void displayInfo(){
        System.out.println("Vehicle's information:");
        System.out.println("- Brand:  " + this.brand + "; Years: " + this.years);
    }
}

class Car extends Vehicle{
    private String model;
    public Car()
    {
        super();
        this.model = "unknow model";
    }
    public Car(String brd, int y, String mod){
        super(brd,y);
        this.model = mod;
    }
    
    @Override
    public void displayInfo(){
        System.out.println("Car's information:");
        System.out.println("- Brand:  " + getBrand() + "; Years: " + getYears() + "; Model: " + this.model);
    }
}

class Bike extends Vehicle{
    private int engineCapacity;
    public Bike()
    {
        super();
        this.engineCapacity = 0;
    }
    public Bike(String brd, int y, int ec){
        super(brd,y);
        this.engineCapacity = ec;
    }
    
    @Override
    public void displayInfo(){
        System.out.println("Bike's information:");
        System.out.println("- Brand:  " + getBrand() + "; Years: " + getYears() + "; Engine Capacity: " + this.engineCapacity);
    }
}

public class exercise_1 {
    public static void main(String[] args) {
        Vehicle c1 = new Car("Honda", 3, "Civic");
        c1.displayInfo();
        Vehicle b1 = new Bike("Yamaha", 2, 150);
        b1.displayInfo();
    }
}

File: LTNC/Exercise/week_2/exercise_1/exercise_1.py
from typing import override #Python 3.12+ for @override decorator

class Vehicle:
    def __init__(self, brd: str = "Unknown", yrs: int = 0):
        self.__brand: str = brd # Private attribute
        self.__year: int = yrs

    def getBrand(self) -> str:
        return self.__brand
    def getYear(self) -> int:
        return self.__year

    def setBrand(self, brd: str)-> None:
        self.__brand = brd
    def setYear(self, yrs: int)-> None:
        self.__year = yrs

    def displayInfo(self) -> None:
        print("Vehicle's Information:")
        print(f"-Brand: {self.__brand}; Year: {self.__year}")

class Car(Vehicle):
    def __init__(self, brd:str = "Unknown", yrs:int = 0, mod:str = "Unknown"):
        super().__init__(brd, yrs)
        self.__model: str = mod
    
    def getModel(self) -> str:
        return self.__model
    def setModel(self, mod: str)-> None:
        self.__model = mod
    
    @override
    def displayInfo(self) -> None:
        print("Car's Information:")
        print(f"-Brand: {super().getBrand()}; Year: {super().getYear()}; Model: {self.__model}")

class Bike(Vehicle):
    def __init__(self, brd:str = "Unknown", yrs:int = 0, engineCapacity:int = 0):
        super().__init__(brd, yrs)
        self.__engineCapacity: int = engineCapacity
    
    def getEngineCapacity(self) -> int:
        return self.__engineCapacity
    def setEngineCapacity(self, engineCapacity: int)-> None:
        self.__engineCapacity = engineCapacity
    
    @override
    def displayInfo(self) -> None:
        print("Bike's Information:")
        print(f"-Brand: {super().getBrand()}; Year: {super().getYear()}; Engine Capacity: {self.__engineCapacity}")

def main():
    print("parameters constructor of Car:")
    car1: Vehicle = Car("Honda", 2021, "Civic")
    car1.displayInfo()

    print("\ndefault constructor of Car:")
    car2: Vehicle = Car()
    car2.displayInfo()
    
    print("\nparameters constructor of Bike:")
    bike1: Vehicle = Bike("Yamaha", 2019, 150)
    bike1.displayInfo()

    print("\ndefault constructor of Bike:")
    bike2: Vehicle = Bike()
    bike2.displayInfo()
    
if __name__ == "__main__":
    main()